src/tools/crossgen2/Common/TypeSystem/Common/Utilities/TypeNameFormatter.cs

```diff
@@ -87,7 +87,29 @@ public abstract class TypeNameFormatter<TState, TOptions>
     {
         public TState AppendName(StringBuilder sb, TypeDesc type, TOptions options)
         {
-            switch (type.Category)
+            TypeFlags category;
+            bool normalCategoryComputation = false;
+            try
```

Comment:
Maybe a virtual GetCategory method (same as the GetContainingType you added below) would be nicer? We could avoid the try/catch here completely (just switch over the type in debug name formatter's override).

Reply:
I don't like that idea as the api would have to return nonsense results in some cases such as Array vs SzArray, and for the DefType case, and I don't like nonsense results to leave a single method.

Reply:
I'd also be willing to remove the entire Category logic from this code and just do the typeswitch, but that will be slower.

Reply:
I think that slower but easier to reason about code in a place that is not perf critical is better.

Reply:
Ok, no worries there. Changed.